server/src/main/java/org/elasticsearch/index/termvectors/TermVectorsService.java

```diff
@@ -321,14 +321,9 @@ private static Fields generateTermVectorsFromDoc(IndexShard indexShard, TermVect
 
     private static ParsedDocument parseDocument(IndexShard indexShard, String index, BytesReference doc,
                                                 XContentType xContentType, String routing) {
-        MapperService mapperService = indexShard.mapperService();
-        DocumentMapperForType docMapper = mapperService.documentMapperWithAutoCreate();
-        ParsedDocument parsedDocument = docMapper.getDocumentMapper().parse(
-                new SourceToParse(index, "_id_for_tv_api", doc, xContentType, routing, Map.of()));
-        if (docMapper.getMapping() != null) {
-            parsedDocument.addDynamicMappingsUpdate(docMapper.getMapping());
-        }
-        return parsedDocument;
+        DocumentMapper documentMapper = indexShard.mapperService().documentMapper();
+        //TODO this throws NPE if there are no mappings yet, we have the same problem in PainlessExecutionAction
```

Comment:
The problem here is that there are cases where some API may want to require parsing documents against an empty index, with no mappings. This is perfectly fine but we don't have a good way to handle this case. Should we make sure that we can use MappingLookup and have a DocumentParser set to it even when empty, but it won't see metadata fields, or shall we reuse DocumentMapper.createEmpty ? Also, we may want to expose a utility method somewhere that creates a parser when needed? It sounds like for this case we may want to enhance DocumentParser and pass it around instead of using it through MappingLookup and DocumentMapper, but I am not terribly sure.

Reply:
If there are no mappings then then we should be returning `null` from `generateTermVectorsForDoc` (because nothing is mapped as text, so there aren't any fields to return).  Maybe we could change the signature of `parseDocument` to take a `DocumentParser` instead of an `IndexShard`, and return early from `generateTermVectorsForDoc` if the shard has no mappings?

Reply:
We could do something similar in PainlessExecuteAction, generating an empty `Document` collection if the document parser is empty (although there I guess we'd want to make `_source` available for scripts)

Reply:
I am not sure I follow: a document is provided from the user and we need to parse it, one way or the other. Maybe we should simulate the situation in which that is the first doc that gets indexed in the index when it is empty. But I am not sure that the parsed document can be empty?

Reply:
I think I would prefer to think about painless execute and term vectors in a follow-up. Do you mind?

Reply:
+1